Fix #2035 Replace glyphicons with material icons

raju249 opened the pull request:
The remaining glyphicons were replaced with equivalent material icons and ``` scripts/pre_commit_linter.py ``` file was also updated for bad pattern checks involving use of glyphicons in future

raju249 commented:
@souravbadami Can you please restart the first two build jobs ?
I don't know what they failed :disappointed:

raju249 commented:
@souravbadami 
Thanks!
All tests passed :smile:

souravbadami commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -37,11 +37,13 @@ <h2 class="oppia-page-heading">Notifications</h2>
-                    class="glyphicon glyphicon-pencil"
-                    title="Exploration commit"></span>
+                    title="Exploration commit">
+                      <i class="material-icons">&#xE150;</i>
```
Incorrect Indentation.

souravbadami commented on core/templates/dev/head/profile/profile.html:
```diff
@@ -128,10 +128,11 @@ <h2 class="oppia-profile-username-small-screen">
         Showing <[currentPageNumber * PAGE_SIZE + 1]> - <[Math.min(numUserPortfolioExplorations, (currentPageNumber + 1) * PAGE_SIZE)]> of <[numUserPortfolioExplorations]>
-        <li class="glyphicon glyphicon-arrow-right" ng-disabled="(currentPageNumber + 1) * PAGE_SIZE >= userEditedExplorations.length"
+        <li ng-disabled="(currentPageNumber + 1) * PAGE_SIZE >= userEditedExplorations.length"
             ng-click="goToNextPage()">
```
Can you pull this up with line 133 ? And use proper indentation for <i> tag below.

souravbadami commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -37,11 +37,13 @@ <h2 class="oppia-page-heading">Notifications</h2>
               <span ng-if="notification.type === 'feedback_thread'"
-                    class="glyphicon glyphicon-comment"
-                    title="Feedback thread"></span>
+                    title="Feedback thread">
```
Maybe this line can go up as we have some space now. Incorrect indentation (line 41).

souravbadami commented on core/templates/dev/head/player/conversation_skin_directive.html:
```diff
@@ -52,8 +52,9 @@
-                <span class="glyphicon glyphicon-play conversation-skin-responses-dropdown-icon"
+                <span class="conversation-skin-responses-dropdown-icon"
                       ng-class="{'conversation-skin-responses-dropdown-icon-rotated': arePreviousResponsesShown}">
+                      <i class="material-icons">&#xE037;</i>
```
Incorrect Indentation.

souravbadami replied:
Incorrect Indentation.

souravbadami commented:
@raju249 -- Reviewed!

souravbadami commented on core/templates/dev/head/dashboard/notifications_dashboard.html:
```diff
@@ -36,12 +36,14 @@ <h2 class="oppia-page-heading">Notifications</h2>
-              <span ng-if="notification.type === 'feedback_thread'"
-                    class="glyphicon glyphicon-comment"
-                    title="Feedback thread"></span>
+              <span ng-if="notification.type === 'feedback_thread'" 
```
Remove trailing whitespace.

raju249 commented:
@souravbadami Can you please re-start the fourth build ?

souravbadami commented on scripts/pre_commit_linter.py:
```diff
@@ -74,6 +74,7 @@
     action='store')
 
 BAD_PATTERNS = {
+
```
Remove this line.

souravbadami replied:
This empty line is still here.

raju249 commented:
@souravbadami PTAL :smile: 
/cc @seanlip

souravbadami commented:
@seanlip -- I've taken a pass. Can you PTAL ?

seanlip commented:
The code looks fine, but @souravbadami, @raju249 -- you really need to replicate and test _every single one_ of these changes manually, especially since this is a UI thing. Additional CSS changes may be needed. I did a spot-check and in all cases found that the icons looked seriously misaligned:

![208701](https://cloud.githubusercontent.com/assets/10575562/16002278/ab6046c0-310b-11e6-9f16-987d285225a9.png)

![208702](https://cloud.githubusercontent.com/assets/10575562/16002277/ab5e3362-310b-11e6-8de7-657acb2582df.png)

@raju249, could you please make the necessary changes, and submit additional commits to this PR to fix the UI issues?

@souravbadami, when you review the PR, please manually test to make sure everything shows up correctly.

Thanks!

souravbadami commented:
@seanlip -- I went through all of them and found that each one of them needs improvement. I'll make sure of this next time. Thanks!

raju249 commented:
@souravbadami @seanlip 
I am extremely sorry !
I just replaced them in html.
I guess each one of then just requires top padding 
What you think @souravbadami ?
/cc @seanlip

raju249 commented:
@souravbadami When I add padding-top to any of the images above sent by @seanlip the icon and text both of them come down instead I expected for just the icon to come down.
I am not so well-versed in front end
Any thoughts @souravbadami ?

souravbadami commented:
@raju249 -- You'll have to try-test and decide which modification gives you the perfect UI result for every case. For example, 
![ab5e3362-310b-11e6-8de7-657acb2582df](https://cloud.githubusercontent.com/assets/11073943/16010041/0a6e378a-319d-11e6-82b3-6590904339f7.png)
The issue with the play button in the above case can be solved by adding an additional class with vertical-align attribute as middle.

raju249 commented:
Thanks @souravbadami 
Can you please help me in getting other icons like what exploration should I load so that I see the associated icons.?
I have found both from the above images(the play button for music and previous responses)

raju249 commented:
@souravbadami 
![oppia](https://cloud.githubusercontent.com/assets/8307266/16035017/85df6b42-3232-11e6-9a0d-50259ece0a86.png)

Thanks! Your suggestion worked.
Now I am ``` grep ``` for other icons and figuring out how to find them on site :sweat_smile:

souravbadami commented:
@raju249 -- I don't think that alone will solve the problem for all the cases. For instance,
![screenshot from 2016-06-14 17 37 25](https://cloud.githubusercontent.com/assets/11073943/16042139/e99b5122-3256-11e6-99b3-dd2b22d219e2.png)
Please check for all the cases and then push the changes. Thanks!

raju249 commented:
Working on it @souravbadami 
I actually don't know what all more links contain material-icons.
Can you please help me with other links to see where icons are misplaced :sweat_smile: ?
Like the one you gave for profile here above. and two of them are fixed (the previous responses icon and play button icon)

raju249 commented:
![icon](https://cloud.githubusercontent.com/assets/8307266/16108970/f545988e-33c3-11e6-9bc5-312f7c8113b1.png)

@souravbadami 
Profile page is done :smile: 
Now I am looking for other pages

raju249 commented:
@souravbadami 
Can you please tell me which link or page contains "+" and "-"  icons ?

seanlip commented:
@raju249 you need to try and find these yourself! Please look at the files you're modifying, and try to figure out where and how they're being used in the application.

raju249 commented:
Sure @seanlip :smile:

souravbadami commented:
@raju249 -- I think, we can use another arrow or a smaller version for this. This is how it looks on the develop branch.
![screenshot from 2016-06-17 12 33 21](https://cloud.githubusercontent.com/assets/11073943/16143154/8746502e-3488-11e6-9da5-22f48d32345c.png)

If you are facing problem in finding the '+' and '-' icons, they are used in the rte collapsible block.

raju249 commented:
@souravbadami 
The one I have used are 24dp .
The lowest I am allowed to go is 18dp on material icons on google's website.
This are the arrows that have this navigation kind of shape other types are like "Less or greater than symbol"  
Should I use arrows of size 18 dp?

souravbadami commented:
@raju249 -- FYI, this is how I figured it out: I went through the diff of your PR and found that the '+' and '-' changes were made in `extensions/rich_text_components/Collapsible/Collapsible.html`.
I'm not sure about the navigation kind. Maybe we can use chevron left and right icons. I think, @seanlip can tell us a better option.

For the icon size, we can use smaller size. Please go through the material-icons size declarations in the css file.

seanlip commented:
For navigation, see what is done with the progress dots in the learner view. It's fine to do the same thing.

raju249 commented:
@souravbadami 
I was actually looking for the url for page that contains "+" and "-" icons.
Can you please check in this commit? :sweat_smile: 
What else is misplaced ?

raju249 commented:
@seanlip 
I didn't get that about the learner view .. :sweat_smile: 
Can you please explain more ?

seanlip commented:
Load the "welcome" exploration from the /admin page. Go to /explore/0 and play it a bit. Look at the progress dots on the top -- there are arrows there too. I was referring to those

raju249 commented:
![untitled](https://cloud.githubusercontent.com/assets/8307266/16145086/b45d2dba-3494-11e6-97ce-8a73f885d8c3.png)

@seanlip I have used exactly the same icons :smiley: 
I reduced the font-size to 18px..but this seems a bit hacky solution 
Is this ok?
@seanlip 
@souravbadami

seanlip commented:
Thanks, @raju249! I defer to @souravbadami's judgement, since he's primary reviewer for this PR. :)

raju249 commented:
ok @seanlip 
So lets see what @souravbadami says for this 18px fix :+1: 
Or else I will switch to chevron arrows :smile:

raju249 commented:
@souravbadami Is 18px size fine?

souravbadami commented on core/templates/dev/head/profile/profile.html:
```diff
@@ -128,11 +128,10 @@ <h2 class="oppia-profile-username-small-screen">
-        <li class="glyphicon glyphicon-arrow-right" ng-disabled="(currentPageNumber + 1) * PAGE_SIZE >= userEditedExplorations.length"
-            ng-click="goToNextPage()">
-        </li>
+        <i class="material-icons" ng-disabled="currentPageNumber == 0"
```
Why <i is used instead of <li ?

raju249 replied:
@souravbadami 
The first image of misplaced arrows that you gave (one arrow was above and one arrow was below the "Showing 6-7" text) was due to the <li> tag.
When I changed it to ``` <i> ``` the arrows and the text were aligned in the same line i.e the lastest image i uploaded was due to using ``` <i> ``` instead of ``` <li> ```

souravbadami replied:
Okay, I'll explain. It was caused because of using `<i>` within `<li>`.

raju249 replied:
Yes :sweat_smile:

souravbadami commented:
@raju249 -- I think, the size is still same, i.e., It's not changed to 18px yet.

raju249 commented:
@souravbadami I didn't push the commit.
Was just waiting for your review on the image I uploaded

souravbadami commented:
Oh! yeah, it looks fine to me. You can push the changes.

raju249 commented:
@souravbadami Done !

souravbadami commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -65,6 +65,14 @@ button.md-button.md-default-theme.oppia-learner-continue-button:hover {
+  vertical-align: middle;
+}
+/* Style for smaller size of material icons for arrows on profile page. */
+.small-size{
```
It is not necessary to define this. We already have md-18 for material icons.

raju249 replied:
ohh So I have to use ``` md-18 ``` class right?

souravbadami replied:
Yes.

souravbadami commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -65,6 +65,14 @@ button.md-button.md-default-theme.oppia-learner-continue-button:hover {
   background-color: #115FD4;
 }
 
+/* Style for unaligned material-icons. */
```
This should be something like, Attributes for material-icons.

raju249 replied:
Yeah sure!
I will edit it online on github and commit them :+1:

souravbadami commented on core/templates/dev/head/profile/profile.html:
```diff
@@ -128,11 +128,10 @@ <h2 class="oppia-profile-username-small-screen">
-        </li>
+        <i class="material-icons md-18" ng-disabled="currentPageNumber == 0"
+           ng-click="goToPreviousPage()">&#xE5C4;</i>
+           Showing <[currentPageNumber * PAGE_SIZE + 1]> - <[Math.min(numUserPortfolioExplorations, (currentPageNumber + 1) * PAGE_SIZE)]> of <[numUserPortfolioExplorations]>
```
Use 2 space indentation.

raju249 replied:
Yes!
@souravbadami I really feel so dumb :sweat_smile: like very minor mistakes that I am commiting.
Is this done by every first time contributor?

souravbadami replied:
Not so much. :) But in the end, you'll learn a lot from your mistakes.

raju249 replied:
Thanks @souravbadami 
I will surely try to minimize my mistakes (like the indentation and extra spaces)
Thanks a lot :-)

souravbadami commented on core/templates/dev/head/profile/profile.html:
```diff
@@ -128,11 +128,10 @@ <h2 class="oppia-profile-username-small-screen">
+        <i class="material-icons md-18" ng-disabled="currentPageNumber == 0"
+           ng-click="goToPreviousPage()">&#xE5C4;</i>
+           Showing <[currentPageNumber * PAGE_SIZE + 1]> - <[Math.min(numUserPortfolioExplorations, (currentPageNumber + 1) * PAGE_SIZE)]> of <[numUserPortfolioExplorations]>
+        <i class="material-icons md-18" ng-disabled="(currentPageNumber + 1) * PAGE_SIZE >= userEditedExplorations.length"   ng-click="goToNextPage()">&#xE5C8;</i>
```
Remove extra space before ng-click.

souravbadami commented on core/templates/dev/head/profile/profile.html:
```diff
@@ -128,11 +128,10 @@ <h2 class="oppia-profile-username-small-screen">
-        </li>
+        <i class="material-icons md-18" ng-disabled="currentPageNumber == 0"
+           ng-click="goToPreviousPage()">&#xE5C4;</i>
+             Showing <[currentPageNumber * PAGE_SIZE + 1]> - <[Math.min(numUserPortfolioExplorations, (currentPageNumber + 1) * PAGE_SIZE)]> of <[numUserPortfolioExplorations]>
```
I think you didn't got me. It'll be 2 space indentation from `<i>`.

raju249 replied:
On line 132 ?

souravbadami replied:
On line 133.

raju249 replied:
@souravbadami There is no ``` <i> ``` 
on line 133.

souravbadami replied:
I meant indentation of 'Showing <...' with respect to `<i>` on line 131.

raju249 replied:
Ohh ..So the indentation of ``` Showing < ``` should be 2 spaces from ``` <i> ``` on line 131.
Right?

souravbadami commented:
@raju249 -- Have you looked into '+' icon issue ? I think it'll be better to include that in your next commit. Thanks!

raju249 commented:
@souravbadami 
I was not able to find the page (the web page) that had "+"  or "-" icon :disappointed:

souravbadami commented:
@raju249 -- It's not a page. It is part of an extension called rich text editor.

raju249 commented:
Oh I see
@souravbadami --How can I see that on oppia?

raju249 commented:
@souravbadami 
I think the "+" and "-" icons are aligned properly.

![untitled](https://cloud.githubusercontent.com/assets/8307266/16169772/1163a2a0-3558-11e6-99eb-9b9a8c2e6a02.png)

All thanks to @seanlip for teaching me how to work like a responsible contributor :smile: 
I found the extension and saw that icons are aligned properly :smile:

souravbadami commented:
@raju249 -- I don't think it is well aligned. Look at the previous version. The size appears to be larger and it's aligned a little lower from the middle.
![screenshot from 2016-06-17 12 35 50](https://cloud.githubusercontent.com/assets/11073943/16171093/53bf44ca-3584-11e6-9a22-152707c66965.png)

raju249 commented:
@souravbadami I think I need to use md-18 ....My icons appear larger

souravbadami commented:
Yes, It needs to be fixed.

raju249 commented:
![untitled](https://cloud.githubusercontent.com/assets/8307266/16171573/fe115cfa-358f-11e6-840d-822a10144008.png)

@souravbadami 
I changed the ``` font-size ``` to ``` 16px ``` and removed the span tags and moved the classes in span tag in <i tags
This is what I got in the image above.
The icons are small and also moved upwards a bit
What you think?

raju249 commented:
![untitled](https://cloud.githubusercontent.com/assets/8307266/16171746/dd8ee59c-3594-11e6-9a2d-134b152413df.png)

I think now its aligned in the middle @souravbadami 
I added ``` padding-top:1px ``` and came down a pixel donwwards.
What you think? from both of the above which is the expected?

souravbadami commented:
@raju249 -- It looks better. But a simple solution will be to just add md-18 to the affected icons. Have you tried that ?

raju249 commented:
Yes @souravbadami ..
But the top and bottom of icons turn flat and look something like two C's joined by two vertical lines at the top and bottom  😅

souravbadami commented:
Okay. Please go ahead with your solution.

raju249 commented:
@souravbadami 
Done!
Thanks.

raju249 commented:
@souravbadami 
Done.
I am sorry that line was still there..may be it was because I edited it online and didn't make the change locally :sweat_smile: 
But its done now .Thanks!

seanlip commented:
@souravbadami @raju249 is this ready for merge?

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -65,6 +65,10 @@ button.md-button.md-default-theme.oppia-learner-continue-button:hover {
 }
 
+/* Attributes for material-icons. */
+.material-icons{
```
Careful. With this you are affecting everything with .material-icons on the site, not just the ones you changed. Is this intentional?

seanlip replied:
(If it is, please check all the other instances of .material-icons visually to make sure there aren't any regressions, and remove any duplicate "vertical-align: middle;" attributes.)

raju249 replied:
@seanlip I checked other icons and are not affected by this class
It only aligns the Play icon in the middle as expected 
Thanks!

seanlip replied:
There are other icons with .material-icons classes in the codebase, and some of them already have "vertical-align: middle;" -- see e.g. the site feedback form link in base.html for an example; there are others. Can you please remove the duplicate inline style from these?

Also, lint issue: please add one space before the {.

Thanks!

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -77,6 +81,10 @@ button.md-button.md-default-theme.oppia-learner-continue-button:hover {
 .material-icons.md-dark {
   color: rgba(0, 0, 0, 0.54);
 }
+.middle{
```
Class name is too vague.

Not sure if you need this though. What happens if you just use md-18? I think you mentioned it looked odd; could you attach a screenshot?

raju249 replied:
![untitled](https://cloud.githubusercontent.com/assets/8307266/16185233/ea372efc-36df-11e6-87b0-9f18c8d14acc.png)

@seanlip  @souravbadami 
This is what it looks like when we use ``` md-18 ```
Browser : Firefox 47.0
Platform : Windows 8.1
Thanks

seanlip replied:
Hm. I think that's fine, actually. Can we leave it like that for now, and get rid of the ".middle" class? There's value in not having too many special cases.

seanlip commented on extensions/rich_text_components/Collapsible/Collapsible.html:
```diff
@@ -2,12 +2,8 @@
-        <span class="pull-right" ng-show="isOpen">
-          <span class="glyphicon glyphicon-minus-sign"></span>
-        </span>
+        <i class="material-icons pull-right middle" ng-hide="isOpen">&#xE147;</i>
```
Drop 'middle' class.

seanlip commented on extensions/rich_text_components/Collapsible/Collapsible.html:
```diff
@@ -2,12 +2,8 @@
-          <span class="glyphicon glyphicon-minus-sign"></span>
-        </span>
+        <i class="material-icons pull-right middle" ng-hide="isOpen">&#xE147;</i>
+        <i class="material-icons new pull-right middle" ng-show="isOpen">&#xE15C;</i>
```
Drop 'middle' and 'new' classes.

seanlip commented:
Hi @raju249, I'm taking over from @souravbadami for a bit since he's having exams at the moment. This is looking pretty good, I just had a few comments!

raju249 commented:
@seanlip 
Done as per your instructions :smile: 
Removed duplicate (I use Sublime editor so I used Find to get all material-icons and removed duplicate)
Removed "middle" and "new" classes.
Thanks

seanlip commented on core/templates/dev/head/profile/profile.html:
```diff
@@ -128,11 +128,10 @@ <h2 class="oppia-profile-username-small-screen">
-        </li>
+        <i class="material-icons md-18" ng-disabled="currentPageNumber == 0"
+           ng-click="goToPreviousPage()">&#xE5C4;</i>
+          Showing <[currentPageNumber * PAGE_SIZE + 1]> - <[Math.min(numUserPortfolioExplorations, (currentPageNumber + 1) * PAGE_SIZE)]> of <[numUserPortfolioExplorations]>
```
Nit: indentation here is wrong, it is at the same level as the i-tags so it should be deindented by 2 spaces.

seanlip commented:
Thanks, @raju249! I only have one tiny nit. Otherwise LGTM!

seanlip commented:
LGTM. Thanks @raju249! Will merge when Travis checks pass.

raju249 commented:
Thanks a lot @seanlip 
Got to learn many things in this issue :smile:
